src/group_theory/quaternion_group.lean

```diff
@@ -0,0 +1,257 @@
+/-
+Copyright (c) 2021 Julian Kuelshammer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Julian Kuelshammer
+-/
+import data.zmod.basic
+import group_theory.order_of_element
+import data.nat.basic
+import tactic.interval_cases
+import group_theory.dihedral
+
+/-!
+# Quaternion Groups
+
+We define the (generalised) quaternion groups `quaternion_group n` of order `4n`, also known as
+dicyclic groups, with elements `a i` and `xa i` for `i : zmod n`. The (generalised) quaternion
+groups can be defined by the presentation `⟨a, x | a²ⁿ = 1, x² = aⁿ, x⁻¹ax=a⁻¹⟩`. We write `a i` for
+`aⁱ` and `xa i` for `xaⁱ`. For `n=2` the quaternion group `quaternion_group 2` is isomorphic to the
+unit integral quaternions `units (quaternion ℤ)`.
```

Comment:
It's a little confusing that you use both `x` and `xa` to refer to the same thing here.

Reply:
I don't understand. `xa` depends on some `i : zmod (2 * n)` and should be read $x*a^i$ in mathematics. The maths $x$ corresponds to the lean `xa 0`.

Reply:
Oh! :-) Sorry, I'm dumb. However you should write the doc-string to account for dumb people... Instead of saying we write `xa i` for `xaⁱ`, you could say we write `xa i` for $$x * aⁱ$$, to be clearer.

Reply:
(I failed to understand that one `xa` was a symbol in Lean, and the other `xa` was mathematicians' notation for multiplying two different symbols...)